src/Radicle/Internal/PrimFns.hs

```diff
@@ -69,12 +70,41 @@ purePrimFns = fromList $ allDocs $
               pure $ List [val, bindingsToRadicle st']
           xs -> throwErrorHere $ WrongNumberOfArgs "base-eval" 2 (length xs)
       )
-    , ( "pure-env"
+    , ( "pure-state"
       , "Returns a pure initial radicle state. This is the state of a radicle\
         \ chain before it has processed any inputs."
       , \case
           [] -> pure $ bindingsToRadicle (pureEnv :: Bindings (PrimFns m))
-          xs -> throwErrorHere $ WrongNumberOfArgs "pure-env" 0 (length xs)
+          xs -> throwErrorHere $ WrongNumberOfArgs "pure-state" 0 (length xs)
+      )
+    , ( "state->env"
+      , "Extract the environment from a radicle state."
+      , oneArg "state->env" $ \case
+          State s -> pure (VEnv (env s))
+          v -> throwErrorHere $ TypeError "state->env" 0 TState v
+      )
+    , ( "set-binding"
```

Comment:
How about making `set-binding` and `get-binding` operate on a state rather than the environment? This would eliminate the need for the `VEnvF` value type and the `state->env` and `set-env` functions and would make it generally nicer to use.

Reply:
We are also inconsitent with names. We use the term `binding` here for what we call environment in the Haskell codebase. There `binding` refers to more stuff like the primitives and handles.

Reply:
> This would eliminate the need for the VEnvF value type

Yes, I thought about this, but there is one place where environments are used standalone: modules. Currently a module is dict:

```clojure
{:module 'foo
 :export ['bar]
 :env <env>}
```

I agree that the naming is inconsistent in the codebase (especially the radicle code which refers to the full-state as `env` a lot). I think the most important is to present clear naming on the radicle side (which is what we expose to users), and this shouldn't be affected by the haskell source (which we can update later). I suggest:

- `state` for the machine-state,
- `env` for the enviornment, part of the `state`,
- a `binding` refers to a particular `(identifier, value)` pair an an `env`.

Reply:
I like the terminology.

Too bad we can’t get rid of `VEnvF`. Maybe we should consider using state for modules in the future.

Reply:
The part of the `state` that is not `env` is the state of the references; I don't think that should be captured as part of a module definition.